### PR TITLE
Implement datasize, dataoffset and datacopy for yul interpreter.

### LINKED ISSUE #6165: yulInterpreter: datacopy call fails the assertion m_functions.count(_funCall.functionName.name) in ExpressionEvaluator
## Description

(depends on #6180 )

The following code snippet

```
{datacopy(8,1,8)}
```

fails this assertion

https://github.com/ethereum/solidity/blob/6b12726f51c25840b3ccb231018fc37e994869e7/test/tools/yulInterpreter/Interpreter.cpp#L191

## Environment

- Compiler version: develop commit `21e7cadad`
- Target EVM version (as per compiler settings): constantinople
- Framework/IDE (e.g. Truffle or Remix): n/a
- EVM execution environment / backend / blockchain client: n/a
- Operating system: n/a

## Steps to Reproduce

```
$ ./test/tools/yulrun --input-file <file.yul>
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<langutil::InternalCompilerError>'
  what():  
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/6165

### CODE REVIEW

File: test/tools/yulInterpreter/Interpreter.h [hunk 1]
```diff
@@ -32,6 +32,8 @@
 
 namespace yul
 {
+class Dialect;
```
Comment (bshastry) [inline test/tools/yulInterpreter/Interpreter.h]: This needs to be a `struct` to match its definition type.

File: test/tools/yulrun.cpp [hunk 2]
```diff
@@ -88,7 +88,8 @@ void interpret(string const& _source)
 	InterpreterState state;
 	state.maxTraceSize = 10000;
 	state.maxMemSize = 0x20000000;
-	Interpreter interpreter(state);
+	shared_ptr<Dialect> dialect(EVMDialect::strictAssemblyForEVMObjects(langutil::EVMVersion{}));
+	Interpreter interpreter(state, *dialect);
```
Comment (bshastry) [inline test/tools/yulrun.cpp:92]: `test/tools/ossfuzz/yulFuzzerCommon.cpp` needs an update to reflect this API change.

Comment (bshastry) [reply test/tools/yulrun.cpp:92]: Also, `test/tools/ossfuzz/yulFuzzerCommon.h` needs a `#include <libyul/backends/evm/EVMDialect.h>`

File: test/tools/yulInterpreter/Interpreter.cpp [hunk 3]
```diff
@@ -24,6 +24,8 @@
 
 #include <libyul/AsmData.h>
 #include <libyul/Utilities.h>
+#include <libyul/Dialect.h>
```
Comment (erak) [inline test/tools/yulInterpreter/Interpreter.cpp:26]: Should be sorted alphabetically.

File: test/tools/yulInterpreter/EVMInstructionInterpreter.cpp [hunk 4]
```diff
@@ -444,6 +444,26 @@ u256 EVMInstructionInterpreter::eval(
...
+	if (_name == "datasize"_yulstring)
+		return u256(keccak256(h256(_arguments.at(0)))) & 0xfff;
+	else if (_name == "dataoffset"_yulstring)
+		return u256(keccak256(h256(_arguments.at(0) + 2))) & 0xfff;
+	else if (_name == "datacopy"_yulstring)
+	{
+		// This is identical to codecopy.
+		if (logMemoryWrite(_arguments.at(0), _arguments.at(2)))
+			copyZeroExtended(
+				m_state.memory, m_state.code,
```
Comment (erak) [inline test/tools/yulInterpreter/EVMInstructionInterpreter.cpp]: I know that it's a little waste of lines, but shouldn't they all have a newline?

Comment (chriseth) [reply test/tools/yulInterpreter/EVMInstructionInterpreter.cpp]: oh yes of course!

**Comment by chriseth** [comment]
Updated - let's see if ossfuzz compiles...

File: test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp [hunk 5]
```diff
@@ -80,6 +80,7 @@ extern "C" int LLVMFuzzerTestOneInput(uint8_t const* _data, size_t _size)
 	{
 		yulFuzzerUtil::interpret(
 			os1,
+			*EVMDialect::strictAssemblyForEVMObjects(langutil::EVMVersion()),
```
Comment (bshastry) [inline test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: Dialect is the third argument going by the definition of interpret. So please change to 

```
		yulFuzzerUtil::interpret(
			os1,
            stack.parserResult()->code,
			*EVMDialect::strictAssemblyForEVMObjects(langutil::EVMVersion()),
...
```

File: test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp [hunk 6]
```diff
@@ -96,6 +97,7 @@ extern "C" int LLVMFuzzerTestOneInput(uint8_t const* _data, size_t _size)
 	{
 		yulFuzzerUtil::interpret(
 			os2,
+			*EVMDialect::strictAssemblyForEVMObjects(langutil::EVMVersion()),
```
Comment (bshastry) [inline test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: same here

File: test/tools/ossfuzz/yulProto_diff_ossfuzz.cpp [hunk 7]
```diff
@@ -76,6 +77,7 @@ DEFINE_PROTO_FUZZER(Function const& _input)
 	{
 		yulFuzzerUtil::interpret(
 			os1,
+			*EVMDialect::strictAssemblyForEVMObjects(langutil::EVMVersion()),
```
Comment (bshastry) [inline test/tools/ossfuzz/yulProto_diff_ossfuzz.cpp]: and here

Comment (bshastry) [reply test/tools/ossfuzz/yulProto_diff_ossfuzz.cpp]: and please change the second interpret call further down in the same source file accordingly.

File: test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp (hunk 5 above)
Comment (chriseth) [reply test/tools/ossfuzz/strictasm_diff_ossfuzz.cpp]: oh shoot!

**Comment by bshastry** [comment]
@chriseth This PR now passes ossfuzz build

**Pull request merged**